### PR TITLE
Support for inhomogeneous conductivities in ClosedCoilSolver

### PR DESCRIPTION
**Pull request by Heinrich-BR**
This is a cleaner and segfault-free implementation of the conductivity coefficient changes for the ClosedCoilSolver with some name changes for variables related to the gradient of the scalar field within ScalarPotentialSource, OpenCoilSolver, and ClosedCoilSolver.

### CODE REVIEW

**Review by alexanderianblair — CHANGES_REQUESTED** [review]
Looks like this is currently missing any test checking that inhomogeneous conductivities result in the correct current distribution when plugged into `OpenCoilSolver` or `ClosedCoilSolver`.

Could also benefit from a few extra comments defining the E field in terms of GradPotential (and thus the choice of sign of this potential w.r.t. the usual electric potential).

File: examples/closed_coil/Main.cpp [hunk 1]
```diff
@@ -2,14 +2,26 @@
 
 const char * DATA_DIR = "../../data/";
 
+static void
+zeroVec(const mfem::Vector & x, mfem::Vector & V)
```
[the remark this answers is not in the document]
Comment (Heinrich-BR) [reply examples/closed_coil/Main.cpp:6]: This was just something that remained from my tests with boundary conditions. I've removed it now

File: examples/closed_coil/Main.cpp [hunk 2]
```diff
@@ -2,14 +2,26 @@
...
+
 hephaestus::Coefficients
 defineCoefficients()
 {
   hephaestus::Coefficients coefficients;
   coefficients.scalars.Register(
       "magnetic_permeability", new mfem::ConstantCoefficient(M_PI * 4.0e-7), true);
   coefficients.scalars.Register(
-      "electrical_conductivity", new mfem::ConstantCoefficient(1.0), true);
+      "electrical_conductivity", new mfem::FunctionCoefficient(sigmafunc), true);
```
[the remark this answers is not in the document]
Comment (Heinrich-BR) [reply examples/closed_coil/Main.cpp:24]: That's fair enough, Just changed that

File: test/unit/test_opencoil.cpp [hunk 3]
```diff
@@ -40,9 +40,9 @@ TEST_CASE("OpenCoilTest", "[CheckData]")
   fespaces.Register(std::string("HCurl"), &HCurlFESpace, false);
 
   hephaestus::GridFunctions gridfunctions;
-  gridfunctions.Register(std::string("J"), &j, false);
+  gridfunctions.Register(std::string("E"), &E, false);
 
-  ocs_params.SetParam("SourceName", std::string("J"));
+  ocs_params.SetParam("GradPotentialName", std::string("E"));
```
[the remark this answers is not in the document]
Comment (Heinrich-BR) [reply test/unit/test_opencoil.cpp:45]: Cool, I'll create one!

File: src/sources/open_coil.hpp [hunk 4]
```diff
@@ -78,23 +78,24 @@ class OpenCoilSolver : public hephaestus::Source
...
   std::string V_gf_name_;
   std::string I_coef_name_;
   std::string cond_coef_name_;
 
   // Parent mesh, FE space, and current
   mfem::ParMesh * mesh_parent_;
-  mfem::ParGridFunction * J_parent_;
-  mfem::ParGridFunction * Jt_parent_;
+  mfem::ParGridFunction * grad_p_parent_;
+  mfem::ParGridFunction * grad_p_t_parent_;
```
[the remark this answers is not in the document]
Comment (Heinrich-BR) [reply src/sources/open_coil.hpp:98]: Done

File: examples/open_coil/Main.cpp [hunk 5]
```diff
@@ -8,20 +8,24 @@ zeroVec(const mfem::Vector & x, mfem::Vector & V)
...
-  hephaestus::Subdomain air("air", 2);
-  air.scalar_coefficients.Register(
-      "electrical_conductivity", new mfem::ConstantCoefficient(1.0), true);
-  hephaestus::Coefficients coefficients(std::vector<hephaestus::Subdomain>({coil, air}));
+  hephaestus::Coefficients coefficients;
   coefficients.scalars.Register(
       "magnetic_permeability", new mfem::ConstantCoefficient(M_PI * 4.0e-7), true);
 
+  // Electrical conductivity
+  coefficients.scalars.Register(
```
[the remark this answers is not in the document]
Comment (Heinrich-BR) [reply examples/open_coil/Main.cpp:26]: I've set it to be a constant coefficient. I've also renames zeroVec in this case to be a constVec, to remove any confusion the name might cause

File: src/sources/closed_coil.hpp [hunk 6]
```diff
@@ -65,35 +65,37 @@ class ClosedCoilSolver : public hephaestus::Source
...
+  mfem::Coefficient * sigma_;
   mfem::Coefficient * Itotal_;
   std::vector<int> old_dom_attrs;
   hephaestus::InputParameters solver_options_;
 
-  // Seting J_transfer_ to true will negatively affect performance, but
-  // the resulting source current will be correct for visualisation purposes.
-  // Only set to true if you wish to view the final current.
-  bool J_transfer_;
+  // Seting grad_phi_transfer_ to true will negatively affect performance, but
```
[the remark this answers is not in the document]
Comment (Heinrich-BR) [reply src/sources/closed_coil.hpp]: I've expanded the comment to discuss the definition of grad_phi and the equation being solved

File: test/unit/test_opencoil.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Heinrich-BR) [reply test/unit/test_opencoil.cpp:45]: Done!

**Review by alexanderianblair — APPROVED** [review]
Thanks - addition of the test for an open coil of inhomogeneous conductivity is particularly useful!
